eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala

```diff
@@ -0,0 +1,194 @@
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.{OutPoint, Transaction}
+import fr.acinq.eclair.NodeParams
+import fr.acinq.eclair.blockchain.CurrentBlockCount
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient
+import fr.acinq.eclair.channel.publish.TxPublisher.TxPublishInfo
+
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+
+/**
+ * This actor publishes a fully signed transaction and monitors its status.
+ * It detects if the transaction is evicted from the mempool, and reports details about its status.
+ */
+object MempoolTxMonitor {
+
+  // @formatter:off
+  sealed trait Command
+  case class Publish(replyTo: ActorRef[TxResult], tx: Transaction, input: OutPoint) extends Command
+  private case object PublishOk extends Command
+  private case class PublishFailed(reason: Throwable) extends Command
+  private case class InputStatus(spentConfirmed: Boolean, spentUnconfirmed: Boolean) extends Command
+  private case class CheckInputFailed(reason: Throwable) extends Command
+  private case class TxConfirmations(count: Int) extends Command
+  private case object TxNotFound extends Command
+  private case class GetTxConfirmationsFailed(reason: Throwable) extends Command
+  private case class WrappedCurrentBlockCount(currentBlockCount: Long) extends Command
+  case object Stop extends Command
+  // @formatter:on
+
+  // @formatter:off
+  sealed trait TxResult
+  case object TxConfirmed extends TxResult
+  case class TxRejected(reason: TxPublisher.TxRejectedReason) extends TxResult
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, loggingInfo: TxPublishInfo): Behavior[Command] = {
+    Behaviors.setup { context =>
+      Behaviors.withMdc(loggingInfo.mdc()) {
+        new MempoolTxMonitor(nodeParams, bitcoinClient, context).start()
+      }
+    }
+  }
+
+}
+
+private class MempoolTxMonitor(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, context: ActorContext[MempoolTxMonitor.Command])(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
+
+  import MempoolTxMonitor._
+
+  private val log = context.log
+
+  def start(): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Publish(replyTo, tx, input) => publish(replyTo, tx, input)
+      case Stop => Behaviors.stopped
+    }
+  }
+
+  def publish(replyTo: ActorRef[TxResult], tx: Transaction, input: OutPoint): Behavior[Command] = {
+    context.pipeToSelf(bitcoinClient.publishTransaction(tx)) {
+      case Success(_) => PublishOk
+      case Failure(reason) => PublishFailed(reason)
+    }
+    Behaviors.receiveMessagePartial {
+      case PublishOk =>
+        log.debug("txid={} was successfully published, waiting for confirmation...", tx.txid)
+        waitForConfirmation(replyTo, tx, input)
+      case PublishFailed(reason) if reason.getMessage.contains("rejecting replacement") =>
```

Comment:
Shouldn't we match on the code? It's not really a problem as long as we have non-reg tests

Reply:
There are a few different cases why a transaction fails to replace an existing mempool transaction:

- it doesn't pay enough fees to replace the current tx
- it adds a new unconfirmed input
- it forces the eviction of too many children

All of these use the same error code `TX_MEMPOOL_POLICY` in `bitcoind`, but other errors share that same error code (such as `mempool min fee not met`). So unfortunately I didn't find a way to only catch these errors reliably...